### ISSUE TITLE
Busser sync is a bit slow

### ISSUE THREAD

**Issue by jordansissel** [test-kitchen/test-kitchen#639]
While learning test-kitchen (newbie, here!), I noticed that the busser sync task is pretty slow.

My observation is that, for each file, it takes quite some time. This shell is given some base64 string roughly: `echo some_base64 | path/to/bin/busser ....` which does the installation.

I would *love* to be able to speed this up. Using a tool I have to measure output timing, I see that each file takes about 1 second:

```
% ruby /Users/jls/projects/experiments/ruby/start-time/linetime.rb kitchen verify
0.512034 -----> Starting Kitchen (v1.3.1)
0.675231 -----> Verifying <whoa-centos-70>...
1.504879        Removing /tmp/busser/suites/rspec
1.819707        Uploading /tmp/busser/suites/rspec/bin/pleaserun (mode=0755)
2.630781        Uploading /tmp/busser/suites/rspec/CHANGELOG.asciidoc (mode=0644)
3.559722        Uploading /tmp/busser/suites/rspec/examples/runit.rb (mode=0644)
4.433532        Uploading /tmp/busser/suites/rspec/Gemfile (mode=0644)
5.396277        Uploading /tmp/busser/suites/rspec/Gemfile.lock (mode=0644)
6.306609        Uploading /tmp/busser/suites/rspec/Guardfile (mode=0644)
7.222477        Uploading /tmp/busser/suites/rspec/lib/pleaserun/cli.rb (mode=0644)
8.039261        Uploading /tmp/busser/suites/rspec/lib/pleaserun/configurable.rb (mode=0644)
8.956496        Uploading /tmp/busser/suites/rspec/lib/pleaserun/detector.rb (mode=0644)
9.774567        Uploading /tmp/busser/suites/rspec/lib/pleaserun/errors.rb (mode=0644)
...
```

The left column is "time since start of program"

I use the `kitchen-sync` gem to improve the uploading of the provisioning materials (puppet/chef stuff), but that doesn't (yet?) work to improve the upload speed of the busser.

---

Proposal:

I believe we can shrink the upload step to be over a single ssh connection.

**Comment by jordansissel** [comment]
Timing a single file:

```
[root@whoa-centos-70-jls-3ep16xu8-macintosh-2 ~]# time echo "IyEvYmluL3NoCgplY2hvICJUaGlzIGpvYiBydW5zIHZpYSB1cHN0YXJ0LCBpbnZva2luZyB1cHN0YXJ0IG5vdy4uLiIKZXhlYyBpbml0Y3RsICQxIHt7bmFtZX19Cg==" | sudo -E /tmp/busser/bin/busser deserialize --destination=`sudo -E /tmp/busser/bin/busser suite path`/rspec/templates/upstart/1.5/init.d.sh --md5sum=d26e8254832f29094561d6a07816792b --perms=0644

real    0m0.776s
user    0m0.672s
sys     0m0.099s
```

**Comment by jordansissel** [comment]
```
% time sudo -E /tmp/busser/bin/busser
real    0m0.347s
user    0m0.303s
sys     0m0.044s
```

Getting to the point where `bin/busser` can do things seems to be about 350ms. Another 350ms to write the file and do stuff

For fairness, this is on a pretty slow machine, but 0.75 seconds per file seems really high. I'll keep digging.

**Comment by fnichol** [comment]
@jordansissel oh my, you're totally right, this code hung around for way too long :smile: 

The good news is that, as of yesterday, the code to upload Busser-related files is updated in master. This is the majority of the work which will be released in 1.4.0, including a new plugin type: a Transport (think SSH, WinRM, etc.). Additionally, Busser is now a Verifier, the fourth new plugin type (which can be completely re-implemented/replaced if need be).

Now, the base Verifier (default implementation which Busser uses) will [upload](https://github.com/test-kitchen/test-kitchen/blob/36cc2927a67d68a19f0dd9e42180754ba2bf87ed/lib/kitchen/verifier/base.rb#L68-L70) via a Transport. I haven't profiled the difference between these 2, but I'm pretty sure that we're going to see a speed bump for SSH (yay).

And, yes, allowing for 1 re-usable SSH connection is exactly what I implemented for the SSH Transport. Assuming these updates make your feedback loop snappier, think I'm okay to close?

Thanks for your feedback and time on this!

**Comment by fnichol** [comment]
While we're not as fast as possible, the 1.4.0 release should go a long way to address the Busser sync slowness. We've let Busser (and all Verifiers) use a Transport plugin to upload files and not relying on the crazy Base64-encoding trick. Additionally, we've added [SSH compression](https://github.com/test-kitchen/test-kitchen/pull/675) in an attempt to speed up uploads to instances that are not located on your workstation.

Now that we have the [Transport plugin](https://github.com/test-kitchen/test-kitchen/blob/master/lib/kitchen/transport/base.rb) abstraction, we can keep working on faster uploads and possibly even one-way syncing via rsync as was done in [kitchen-sync](https://github.com/coderanger/kitchen-sync).

So, hopefully we're good to close this out knowing that we'll continue the trek to faster uploads :smile:
